perceval/client.py

```diff
@@ -0,0 +1,185 @@
+    def fetch(self, url, payload=None, headers=None, method=GET,
+              stream=False,
+              retry_on_status_codes=None,
+              retry_after_status_code=None):
+        yield self._fetch(url, payload, headers, method, stream,
+                          retry_on_status_codes,
+                          retry_after_status_code)
+
+    def _fetch(self, url, payload=None, headers=None, method=GET,
+               stream=False,
+               retry_on_status_codes=None,
+               retry_after_status_codes=None):
+
+        self.set_retry_after_status_codes(retry_after_status_codes)
+        self.sleep_for_rate_limit()
+
+        session = self.session or requests.Session()
+        retries = Retry(total=self.MAX_RETRIES,
+                        connect=self.MAX_RETRIES_ON_CONNECT,
+                        read=self.MAX_RETRIES_ON_READ,
+                        redirect=self.MAX_RETRIES_ON_REDIRECT,
+                        status=self.MAX_RETRIES_ON_STATUS,
+                        method_whitelist=False,
+                        status_forcelist=retry_on_status_codes,
+                        backoff_factor=self.DEFAULT_SLEEP_TIME,
+                        raise_on_redirect=True,
+                        raise_on_status=True,
+                        respect_retry_after_header=True)
+        session.mount('http://', HTTPAdapter(max_retries=retries))
+        session.mount('https://', HTTPAdapter(max_retries=retries))
+
+        response = self.__send_request(session, url, payload, headers, method, stream)
+        self.update_rate_limit(response)
+
+        if not self.session:
+            self.close_http_session(session)
+
+        return response
+
+    def sleep_for_rate_limit(self):
+        if self.rate_limit is not None and self.rate_limit <= self.min_rate_to_sleep:
+            seconds_to_reset = self.rate_limit_reset_ts - int(time.time()) + 1
+            cause = "Rate limit exhausted."
+            if self.sleep_for_rate:
+                logger.info("%s Waiting %i secs for rate limit reset.", cause, seconds_to_reset)
+                time.sleep(seconds_to_reset)
+            else:
+                raise RateLimitError(cause=cause, seconds_to_reset=seconds_to_reset)
+
+    def update_rate_limit(self, response):
+        if self.RATE_LIMIT_HEADER in response.headers:
+            self.rate_limit = int(response.headers[self.RATE_LIMIT_HEADER])
+            self.rate_limit_reset_ts = int(response.headers[self.RATE_LIMIT_RESET_HEADER])
+            logger.debug("Rate limit: %s", self.rate_limit)
+        else:
+            self.rate_limit = None
+            self.rate_limit_reset_ts = None
+
+    def __send_request(self, session, url, payload=None, headers=None, method=GET, stream=False):
+        if method == self.GET:
+            response = session.get(url, params=payload, headers=headers, stream=stream)
+        else:
+            response = self.session.post(url, data=payload, headers=headers, stream=stream)
```

Comment:
I think this should be either `session` or `self.session`, but the same on the two calls. Moreover, the code should be something like this:

```
if method == self.GET:
    freq = session.get
else:
    freq = session.post

response = freq(url,  data=payload, headers=headers, stream=stream)
```

Reply:
get methods should pass the payload by the argument **params**, while post methods by the argument **data**. Btw, I guess the behavior is the same, maybe a couple of tests will need to be changed (If I remember well, Bugzilla or Bugzillarest).

Reply:
Yep, you're right. Btw, what's `stream` parameter and why do we need it?

Reply:
The stream parameter is used in **pipermail** and **hyperkitty** clients. It defers downloading the response body until you access the **Response.content** attribute (the default behavior is to immediately download the body of the response).

Reply:
Great. I forgot about it!